safekeeper/src/receive_wal.rs

```diff
@@ -2,204 +2,285 @@
+impl SafekeeperPostgresHandler {
+    /// Wrapper around handle_start_wal_push_guts handling result. Error is
+    /// handled here while we're still in walreceiver ttid span; with API
+    /// extension, this can probably be moved into postgres_backend.
+    pub async fn handle_start_wal_push(
+        &mut self,
+        pgb: &mut PostgresBackend,
+    ) -> Result<(), QueryError> {
+        if let Err(end) = self.handle_start_wal_push_guts(pgb).await {
+            // Log the result and probably send it to the client, closing the stream.
+            pgb.handle_copy_stream_end(end).await;
         }
-    }
-
-    // Send message to the postgres
-    fn write_msg(&mut self, msg: &AcceptorProposerMessage) -> anyhow::Result<()> {
-        let mut buf = BytesMut::with_capacity(128);
-        msg.serialize(&mut buf)?;
-        self.pg_backend.write_message(&BeMessage::CopyData(&buf))?;
         Ok(())
     }
 
-    /// Receive WAL from wal_proposer
-    pub fn run(&mut self, spg: &mut SafekeeperPostgresHandler) -> Result<(), QueryError> {
-        let _enter = info_span!("WAL acceptor", ttid = %spg.ttid).entered();
-
+    pub async fn handle_start_wal_push_guts(
+        &mut self,
+        pgb: &mut PostgresBackend,
+    ) -> Result<(), CopyStreamHandlerEnd> {
         // Notify the libpq client that it's allowed to send `CopyData` messages
-        self.pg_backend
-            .write_message(&BeMessage::CopyBothResponse)?;
-
-        let r = self
-            .pg_backend
-            .take_stream_in()
-            .ok_or_else(|| anyhow!("failed to take read stream from pgbackend"))?;
-        let mut poll_reader = ProposerPollStream::new(r)?;
-
-        // Receive information about server
-        let next_msg = poll_reader.recv_msg()?;
-        let tli = match next_msg {
-            ProposerAcceptorMessage::Greeting(ref greeting) => {
-                info!(
-                    "start handshake with walproposer {} sysid {} timeline {}",
-                    self.peer_addr, greeting.system_id, greeting.tli,
-                );
-                let server_info = ServerInfo {
-                    pg_version: greeting.pg_version,
-                    system_id: greeting.system_id,
-                    wal_seg_size: greeting.wal_seg_size,
-                };
-                GlobalTimelines::create(spg.ttid, server_info, Lsn::INVALID, Lsn::INVALID)?
-            }
-            _ => {
-                return Err(QueryError::Other(anyhow::anyhow!(
-                    "unexpected message {next_msg:?} instead of greeting"
-                )))
-            }
+        pgb.write_message_flush(&BeMessage::CopyBothResponse)
+            .await?;
+
+        // Experiments confirm that doing network IO in one (this) thread and
+        // processing with disc IO in another significantly improves
+        // performance; we spawn off WalAcceptor thread for message processing
+        // to this end.
+        let (msg_tx, msg_rx) = channel(MSG_QUEUE_SIZE);
```

Comment:
Not sure I understand why we need both split workaround and these channels. You can select on pgb.read_message() and reply_rx.recv(). This way you dont need split since you own the stream. If you received the message from the socket you forward it to separate thread. If you received reply message from WalAcceptor you write it to the same instance of pg backend. 

Am I misunderstanding it?

Reply:
Proposed schema is vulnerable to deadlock where both threads are blocked trying to send to the channel.

Reply:
try_send is not good, it should be retried, and when? busy loop or sleep? We anyway need to send at some point, and do that before attempting to get next message, otherwise the thing is pointless, you just accumulate messages in memory. You could imagine some state machine conditionally choosing what to do on each loop, but it doesn't look simpler.

> Additionally what do you think about using an unbounded channel? 

That's bad as memory consumption grows indefinitely. I don't think monitoring that with prometheus really solves this.

> E g [here](https://tomaka.medium.com/a-look-back-at-asynchronous-rust-d54d63934a1c) in paragraph "Flow control is hard" discusses the topic

Indeed flow control is hard. select! here specifically addresses the deadlock mentioned in that link.

Reply:
> try_send is not good,

If you cannot send you simply block on recv from reply channel as I explained above. When you received the message

> That's bad as memory consumption grows indefinitely. I don't think monitoring that with prometheus really solves this.

It will grow a bit, but shouldnt go really hi because you essentially it is a form of request response pattern.  If you cannot process requests fast enough you better know about it. And client shouldnt send too many data without receiving an ack from the other side.

> Indeed flow control is hard. select! here specifically addresses the deadlock mentioned in that link.

Select unnecessary overcomplicates control flow here because it has far richer semantics than try_join_all.

Reply:
> If you cannot send you simply block on recv from reply channel 

It can still be blocked after reply returns something. This is not a trivial thing, a complete description is needed to talk about that. Anyway, this is out of this patch scope as I and Arthur commented below.

> It will grow a bit, but shouldnt go really

Is is not generally bounded by anything, and this is not a request-response, it is bidirectional stream.

> Select unnecessary overcomplicates control flow here

`try_join_all` has different semantics not fitting here, as said elsewhere, we need `select!` exactly. Neither I generally consider `select!` being more complicated that `try_join_all`.

Reply:
>  Anyway, this is out of this patch scope as I and Arthur commented below.

As I get it this change is the main motivation for the patch. It is the motivation for all the squats around split and Framed. So IMO it is exactly in scope of this patch. 

> It can still be blocked after reply returns something

And thats fine, try_send it again on next iteration. 

> try_join_all has different semantics not fitting here

Replied to the relevant thread

>  request-response, it is bidirectional stream.

It is, if I understand it correctly the stream is a essentially a request response with the difference that a bunch or requests can be batched, right? Can there be a huge amount of messages without reply from the other side? From my understanding no, but I may be wrong

Reply:
On which iteration? I don't follow. Should we select in the meanwhile on reading more data? If yes, we pointlessly and unboundedly accumulate data in memory, as I said. If no, it means complicated state machine. And how we read more data without split if reply (out) buffer can be also full, i.e. flushing is WOULDBLOCK? As I said, full description is required to talk about that. Generally, it is not hard to observe that in arbitrary bidi stream (and this is the case here) at least one side must poll reading and writing concurrently, otherwise deadlock risk is inevitable. And I'd prefer to not make any assumptions on how the other side behaves, so better make *both* sides do concurrent read writes.

Reply:
> Can there be a huge amount of messages without reply from the other side? 

Yes, there is generally no constraints. And adding them only to justify poorly written protocol implementation doesn't sound good.